pkg/operator/ceph/cluster/cluster.go

```diff
@@ -57,33 +57,35 @@ const (
 var telemetryMutex sync.Mutex
 
 type cluster struct {
-	ClusterInfo        *client.ClusterInfo
-	context            *clusterd.Context
-	Namespace          string
-	Spec               *cephv1.ClusterSpec
-	clusterMetadata    metav1.ObjectMeta
-	namespacedName     types.NamespacedName
-	mons               *mon.Cluster
-	ownerInfo          *k8sutil.OwnerInfo
-	isUpgrade          bool
-	monitoringRoutines map[string]*controller.ClusterHealth
-	observedGeneration int64
+	ClusterInfo         *client.ClusterInfo
+	context             *clusterd.Context
+	Namespace           string
+	Spec                *cephv1.ClusterSpec
+	clusterMetadata     metav1.ObjectMeta
+	namespacedName      types.NamespacedName
+	mons                *mon.Cluster
+	ownerInfo           *k8sutil.OwnerInfo
+	isUpgrade           bool
+	monitoringRoutines  map[string]*controller.ClusterHealth
+	clusterKeysToRotate ClusterKeysToRotate
+	observedGeneration  int64
 }
 
 func newCluster(ctx context.Context, c *cephv1.CephCluster, context *clusterd.Context, ownerInfo *k8sutil.OwnerInfo) *cluster {
 	return &cluster{
 		// at this phase of the cluster creation process, the identity components of the cluster are
 		// not yet established. we reserve this struct which is filled in as soon as the cluster's
 		// identity can be established.
-		ClusterInfo:        client.AdminClusterInfo(ctx, c.Namespace, c.Name),
-		Namespace:          c.Namespace,
-		Spec:               &c.Spec,
-		clusterMetadata:    c.ObjectMeta,
-		context:            context,
-		namespacedName:     types.NamespacedName{Namespace: c.Namespace, Name: c.Name},
-		monitoringRoutines: make(map[string]*controller.ClusterHealth),
-		ownerInfo:          ownerInfo,
-		mons:               mon.New(ctx, context, c.Namespace, c.Spec, ownerInfo),
+		ClusterInfo:         client.AdminClusterInfo(ctx, c.Namespace, c.Name),
+		Namespace:           c.Namespace,
+		Spec:                &c.Spec,
+		clusterMetadata:     c.ObjectMeta,
+		context:             context,
+		namespacedName:      types.NamespacedName{Namespace: c.Namespace, Name: c.Name},
+		monitoringRoutines:  make(map[string]*controller.ClusterHealth),
+		ownerInfo:           ownerInfo,
+		mons:                mon.New(ctx, context, c.Namespace, c.Spec, ownerInfo),
+		clusterKeysToRotate: ClusterKeysToRotate{},
```

Comment:
I'd like to propose that we not keep track of this at the whole CephCluster level like this.

I think it will be best if each subsection of the CephCluster reconcile checks its own rotation logic. This will be especially important for OSDs because of how differently they are managed. CSI keys also have a much different handling.

Since we have to delegate `shouldRotate` handling for 2 of the subsections, I think it would be best to delegate all of them, to keep consistency.